File: SSO연동가이드/samples/SsoClient.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace Innotium.Sso;

/// <summary>
/// 이노티움 SSO 중앙 집중형 인증 클라이언트 (.NET 6+).
/// 사용자 로그인 검증 + 비밀번호 변경 + 사용자/부서 조회.
/// </summary>
public class SsoClient : IDisposable
{
    private readonly string _baseUrl;
    private readonly string _clientId;
    private readonly string _clientSecret;
    private readonly HttpClient _http;
    private readonly SemaphoreSlim _lock = new(1, 1);

    private string? _token;
    private DateTime _expiresAt = DateTime.MinValue;

    public SsoClient(string baseUrl, string clientId, string clientSecret, HttpClient? http = null)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _clientId = clientId; _clientSecret = clientSecret;
        _http = http ?? new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
    }

    private async Task<string> TokenAsync()
    {
        if (_token != null && DateTime.UtcNow < _expiresAt) return _token;
        await _lock.WaitAsync();
        try
        {
            if (_token != null && DateTime.UtcNow < _expiresAt) return _token;
            var basic = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{_clientId}:{_clientSecret}"));
            using var req = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/apie/sso/oauth/client-token");
            req.Headers.Authorization = new AuthenticationHeaderValue("Basic", basic);
            using var res = await _http.SendAsync(req);
            var body = await ParseAsync(res);
            _token = body.GetProperty("access_token").GetString();
            var expiresIn = body.GetProperty("expires_in").GetInt32();
            _expiresAt = DateTime.UtcNow.AddSeconds(expiresIn - 60);
            return _token!;
        }
        finally { _lock.Release(); }
    }

    // ───── 로그인 검증 ─────
    public async Task<JsonElement> LoginAsync(string loginId, string password)
    {
        return await PostAsync("/apie/sso/auth/login", new { loginId, password });
    }

    // ───── 비밀번호 변경 ─────
    public async Task<JsonElement> ChangePasswordAsync(string loginId, string currentPassword, string newPassword)
    {
        return await PostAsync("/apie/sso/auth/password", new { loginId, currentPassword, newPassword });
    }

    // ───── 조회 ─────
    public async Task<JsonElement> GetUserAsync(long userId)
        => await GetAsync($"/apie/sso/users/{userId}");

    public async Task<JsonElement> ListUsersAsync(string? keyword = null, long? departmentId = null,
                                                    string? status = null, int startIndex = 0, int pageSize = 50)
    {
        var url = $"/apie/sso/users?startIndex={startIndex}&pageSize={pageSize}";
        if (keyword != null) url += "&keyword=" + Uri.EscapeDataString(keyword);
        if (departmentId != null) url += "&departmentId=" + departmentId;
        if (status != null) url += "&status=" + status;
        return await GetAsync(url);
    }

    public async Task<JsonElement> GetDepartmentAsync(long departmentId)
        => await GetAsync($"/apie/sso/departments/{departmentId}");

    public async Task<JsonElement> ListDepartmentsAsync(string? keyword = null, string? status = null,
                                                         int startIndex = 0, int pageSize = 500)
    {
        var url = $"/apie/sso/departments?startIndex={startIndex}&pageSize={pageSize}";
        if (keyword != null) url += "&keyword=" + Uri.EscapeDataString(keyword);
        if (status != null) url += "&status=" + status;
        return await GetAsync(url);
    }

    // ───── 내부 ─────
    private async Task<JsonElement> PostAsync(string path, object body)
    {
        var token = await TokenAsync();
        using var req = new HttpRequestMessage(HttpMethod.Post, _baseUrl + path) { Content = JsonContent.Create(body) };
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        using var res = await _http.SendAsync(req);
        return await ParseAsync(res);
    }

    private async Task<JsonElement> GetAsync(string path)
    {
        var token = await TokenAsync();
        using var req = new HttpRequestMessage(HttpMethod.Get, _baseUrl + path);
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        using var res = await _http.SendAsync(req);
        return await ParseAsync(res);
    }

    private static async Task<JsonElement> ParseAsync(HttpResponseMessage res)
    {
        var text = await res.Content.ReadAsStringAsync();
        var doc = string.IsNullOrWhiteSpace(text) ? JsonDocument.Parse("{}") : JsonDocument.Parse(text);
        if (!res.IsSuccessStatusCode)
        {
            var key = doc.RootElement.TryGetProperty("messageKey", out var mk) ? mk.GetString() : "UNKNOWN";
            throw new SsoApiException((int)res.StatusCode, key ?? "UNKNOWN", doc.RootElement.Clone());
        }
        return doc.RootElement.Clone();
    }

    public void Dispose() => _http.Dispose();
}

public class SsoApiException : Exception
{
    public int StatusCode { get; }
    public string MessageKey { get; }
    public JsonElement Body { get; }
    public SsoApiException(int sc, string mk, JsonElement body)
        : base($"SSO API {sc} {mk}")
    { StatusCode = sc; MessageKey = mk; Body = body; }
}

// 사용 예:
//   using var sso = new SsoClient("https://sso.innotium.com",
//       Environment.GetEnvironmentVariable("SSO_CLIENT_ID")!,
//       Environment.GetEnvironmentVariable("SSO_CLIENT_SECRET")!);
//   var user = await sso.LoginAsync("kim", "plain");
//   long userId = user.GetProperty("userId").GetInt64();
//   await sso.ChangePasswordAsync("kim", "old", "New1!");

File: SSO연동가이드/samples/sso_client.py
"""
이노티움 SSO 중앙 집중형 인증 클라이언트 (Python 3.9+)

기능:
  - JWT 자동 발급/갱신 (client credentials)
  - 로그인 검증 (/apie/sso/auth/login)
  - 비밀번호 변경 (/apie/sso/auth/password)
  - 사용자·부서 조회

Usage:
    from sso_client import SsoClient, SsoApiError

    sso = SsoClient("https://sso.innotium.com",
                    client_id="sso-xxx", client_secret="xxx")

    # 로그인 검증
    try:
        info = sso.login("kim", "plain_password")
        user_id = info["userId"]  # 자체 DB에 FK로 저장
        dept_id = info["primaryDepartment"]["departmentId"]
    except SsoApiError as e:
        if e.message_key == "INVALID_CREDENTIALS":
            print("비밀번호가 틀렸습니다.")

    # 비밀번호 변경
    sso.change_password("kim", "old", "NewPass1234!")

    # 사용자 상세
    user = sso.get_user(user_id)
"""
import time
import base64
import logging
from typing import Optional, Dict, Any, List

import requests

log = logging.getLogger(__name__)


class SsoApiError(Exception):
    def __init__(self, status_code: int, message_key: str, raw: dict):
        self.status_code = status_code
        self.message_key = message_key
        self.raw = raw
        super().__init__(f"SSO API {status_code} {message_key}")


class SsoClient:
    def __init__(self, base_url: str, client_id: str, client_secret: str, timeout: int = 10):
        self.base_url = base_url.rstrip("/")
        self.client_id = client_id
        self.client_secret = client_secret
        self.timeout = timeout
        self._token: Optional[str] = None
        self._expires_at: float = 0.0

    # ──────── 토큰 관리 ────────
    def _fetch_token(self) -> None:
        basic = base64.b64encode(f"{self.client_id}:{self.client_secret}".encode()).decode()
        res = requests.post(
            f"{self.base_url}/apie/sso/oauth/client-token",
            headers={"Authorization": f"Basic {basic}"},
            timeout=self.timeout,
        )
        self._raise_for_error(res)
        body = res.json()
        self._token = body["access_token"]
        self._expires_at = time.time() + body["expires_in"] - 60

    def _auth(self) -> Dict[str, str]:
        if self._token is None or time.time() >= self._expires_at:
            self._fetch_token()
        return {"Authorization": f"Bearer {self._token}"}

    # ──────── 로그인 검증 ────────
    def login(self, login_id: str, password: str) -> Dict[str, Any]:
        """
        외부 앱 로그인 폼에서 호출.
        성공 시 userId, name, departments 등 반환 (ResExtUser).
        실패 시 SsoApiError (INVALID_CREDENTIALS, ACCOUNT_LOCKED 등).
        """
        res = requests.post(
            f"{self.base_url}/apie/sso/auth/login",
            headers={**self._auth(), "Content-Type": "application/json"},
            json={"loginId": login_id, "password": password},
            timeout=self.timeout,
        )
        self._raise_for_error(res)
        return res.json()

    # ──────── 비밀번호 변경 ────────
    def change_password(self, login_id: str, current_password: str, new_password: str) -> bool:
        res = requests.post(
            f"{self.base_url}/apie/sso/auth/password",
            headers={**self._auth(), "Content-Type": "application/json"},
            json={"loginId": login_id,
                  "currentPassword": current_password,
                  "newPassword": new_password},
            timeout=self.timeout,
        )
        self._raise_for_error(res)
        return res.json().get("changed") is True

    # ──────── 사용자 조회 ────────
    def get_user(self, user_id: int) -> Dict[str, Any]:
        res = requests.get(
            f"{self.base_url}/apie/sso/users/{user_id}",
            headers=self._auth(), timeout=self.timeout)
        self._raise_for_error(res)
        return res.json()

    def list_users(self, keyword: Optional[str] = None, department_id: Optional[int] = None,
                    status: Optional[str] = None, start_index: int = 0, page_size: int = 50) -> List[Dict[str, Any]]:
        params = {"startIndex": start_index, "pageSize": page_size}
        if keyword: params["keyword"] = keyword
        if department_id: params["departmentId"] = department_id
        if status: params["status"] = status
        res = requests.get(f"{self.base_url}/apie/sso/users",
                           headers=self._auth(), params=params, timeout=self.timeout)
        self._raise_for_error(res)
        return res.json()

    # ──────── 부서 조회 ────────
    def get_department(self, department_id: int) -> Dict[str, Any]:
        res = requests.get(f"{self.base_url}/apie/sso/departments/{department_id}",
                           headers=self._auth(), timeout=self.timeout)
        self._raise_for_error(res)
        return res.json()

    def list_departments(self, keyword: Optional[str] = None, status: Optional[str] = None,
                          start_index: int = 0, page_size: int = 500) -> List[Dict[str, Any]]:
        params = {"startIndex": start_index, "pageSize": page_size}
        if keyword: params["keyword"] = keyword
        if status: params["status"] = status
        res = requests.get(f"{self.base_url}/apie/sso/departments",
                           headers=self._auth(), params=params, timeout=self.timeout)
        self._raise_for_error(res)
        return res.json()

    # ──────── 유틸 ────────
    @staticmethod
    def _raise_for_error(res: requests.Response) -> None:
        if res.status_code >= 400:
            try:
                body = res.json()
                raise SsoApiError(res.status_code, body.get("messageKey", "UNKNOWN"), body)
            except ValueError:
                raise SsoApiError(res.status_code, "PARSE_ERROR", {"raw": res.text})


# ──────── 실행 예시 ────────
if __name__ == "__main__":
    import os, sys
    logging.basicConfig(level=logging.INFO)

    sso = SsoClient(
        base_url=os.environ.get("SSO_BASE_URL", "https://sso.innotium.com"),
        client_id=os.environ["SSO_CLIENT_ID"],
        client_secret=os.environ["SSO_CLIENT_SECRET"],
    )

    if len(sys.argv) >= 3:
        login_id, password = sys.argv[1], sys.argv[2]
        try:
            info = sso.login(login_id, password)
            print(f"✅ 로그인 성공: {info['name']} (userId={info['userId']})")
            print(f"   소속: {info['primaryDepartment']['departmentName']}")
            print(f"   전체 부서: {[d['departmentName'] for d in info['departments']]}")
        except SsoApiError as e:
            print(f"❌ {e.message_key}")
